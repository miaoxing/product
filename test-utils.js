/**
 * @experimental 可能移到其他位置
 */
import _ from 'lodash';
import {Ret} from 'miaoxing';

export const createSingleSkuProduct = (product) => {
  return _.merge({
    id: 1,
    name: 'alias commodi ipsam',
    intro: 'consequatur officia sequi',
    minPrice: '9',
    minMarketPrice: '10',
    minScore: 0,
    stockNum: 21,
    soldNum: 0,
    image: 'https://dev.test.com/uploads/1/210610/002439539409.png',
    startAt: '2021-06-24 11:17:00',
    endAt: null,
    maxOrderQuantity: 0,
    isAllowAddCart: true,
    isAllowCoupon: true,
    isRequireAddress: true,
    isAllowComment: true,
    deletedAt: null,
    images: [
      {
        url: 'https://dev.test.com/uploads/1/210610/002439539409.png',
        description: '',
      },
      {
        url: 'https://dev.test.com/uploads/1/210610/000924427937.jpg',
        description: '',
      },
      {
        url: 'https://dev.test.com/uploads/1/210610/000918792585.jpg',
        description: '',
      },
      {
        url: 'https://dev.test.com/uploads/1/210610/002904468645.png',
        description: '',
      },
    ],
    detail: {
      content: '<p><img src="https://dev.test.com/uploads/images/20210528/1622199614338752.png" alt="14 - 3 (2).png"/></p><p>Alice: he had a cons<span style="color: #548DD4;">ultation about this, and after a few minutes it puffed away without speaking, but at any r</span>ate,&#39; sai<span style="background-color: #F79646;">d Alice: &#39;allo</span>w me to introduce it.&#39; &#39;I don&#39;t know what a long way. So they got.</p>',
    },
    spec: {
      isDefault: false,
      specs: [
        {
          id: 1,
          name: '默认规格',
          values: [
            {
              id: 2,
              name: '默认值',
            },
          ],
        },
      ],
    },
    skus: [
      {
        id: 1,
        specValueIds: [2],
        stockNum: 21,
      },
    ],
    configs: {
      unit: '件',
    },
    createCartOrOrder: Ret.suc({
      createCart: Ret.suc(),
      createOrder: Ret.suc(),
    }),
  }, product);
};

export const createProduct = (product) => {
  return {
    ...createSingleSkuProduct(product),
    spec: {
      specs: [
        {
          id: 1,
          name: '尺寸',
          values: [
            {id: 1, name: 'S'},
            {id: 2, name: 'M'},
          ],
        },
        {
          id: 2,
          name: '颜色',
          values: [
            {id: 3, name: '红色'},
            {id: 4, name: '蓝色'},
          ],
        },
      ],
    },
    skus: [
      {
        id: 1,
        specValueIds: [1, 3],
        price: '9',
        marketPrice: '10',
        score: 0,
        stockNum: 0,
        soldNum: 0,
        image: '',
      },
      {
        id: 2,
        specValueIds: [1, 4],
        price: '11',
        marketPrice: '12',
        score: 0,
        stockNum: 6,
        soldNum: 0,
        image: '',
      },
      {
        id: 3,
        specValueIds: [2, 3],
        price: '12',
        marketPrice: '13',
        score: 0,
        stockNum: 7,
        soldNum: 0,
        image: '',
      },
      {
        id: 4,
        specValueIds: [2, 4],
        price: '14',
        marketPrice: '15',
        score: 0,
        stockNum: 8,
        soldNum: 0,
        image: '',
      },
    ],
  };
};

