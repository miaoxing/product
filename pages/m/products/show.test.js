import Show from './show';
import {render, waitFor, fireEvent} from '@testing-library/react';
import $, {Ret} from 'miaoxing';
import {createPromise, bootstrap, setUrl, resetUrl} from '@mxjs/test';
import Taro from '@tarojs/taro';
import {createProduct} from '@miaoxing/product/test-utils';

bootstrap();

describe('Show', () => {
  beforeEach(() => {
    setUrl('/products/show?id=1');
  });

  afterEach(() => {
    resetUrl();
  });

  test('basic', async () => {
    const product = createProduct();
    const promise = createPromise();
    const promise2 = createPromise();
    const promise3 = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: createProduct(),
        }),
      }))
      .mockImplementationOnce(() => promise2.resolve({
        ret: Ret.suc({
          data: {
            count: 1,
          },
        }),
      }))
      .mockImplementationOnce(() => promise3.resolve({
        ret: Ret.suc({
          data: {
            isFreeShipping: true,
            rules: [],
          },
        }),
      }));

    const {container, getByText} = render(<Show/>);

    await waitFor(() => {
      expect(getByText('包邮')).not.toBeNull();
    });

    expect(getByText(product.name)).not.toBeNull();

    expect(container).toMatchSnapshot();
    expect($.http).toMatchSnapshot();
  });

  test('action', async () => {
    const promise = createPromise();
    const promise2 = createPromise();
    const promise3 = createPromise();
    const promise4 = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: createProduct(),
        }),
      }))
      .mockImplementationOnce(() => promise2.resolve({
        ret: Ret.suc({
          data: {
            count: 97,
          },
        }),
      }))
      .mockImplementationOnce(() => promise3.resolve({
        ret: Ret.suc({
          data: {
            isFreeShipping: true,
            rules: [],
          },
        }),
      }))
      .mockImplementationOnce(() => promise4.resolve({
        ret: Ret.suc({
          exists: false,
        }),
      }));

    const {container, getByText, findByText, findAllByText} = render(<Show/>);

    await waitFor(() => {
      expect(getByText('包邮')).not.toBeNull();
    });

    // 点击查看图片
    const image = container.querySelector('taro-image-core');

    Taro.previewImage = jest.fn();

    fireEvent.click(image);
    expect(Taro.previewImage).toBeCalled();
    expect(Taro.previewImage.mock.calls).toMatchSnapshot();

    // 点击选择规格
    const selected = getByText('尺寸 / 颜色');
    fireEvent.click(selected);

    const skuValue1 = await findByText('M');
    fireEvent.click(skuValue1);

    const skuValue2 = await findByText('红色');
    fireEvent.click(skuValue2);

    // 加入购物车
    const createCarts = await findAllByText('加入购物车');
    fireEvent.click(createCarts[1]);

    await promise4;
    expect($.http).toMatchSnapshot();

    // 关闭后显示选中的规格
    await findByText('已选：M / 红色');

    // 购物车数量增加
    await findByText('98');
  });
});
