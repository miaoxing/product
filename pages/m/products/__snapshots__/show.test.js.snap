// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show action 1`] = `
Array [
  Array [
    Object {
      "current": "https://dev.test.com/uploads/1/210610/002439539409.png",
      "urls": Array [
        "https://dev.test.com/uploads/1/210610/002439539409.png",
        "https://dev.test.com/uploads/1/210610/000924427937.jpg",
        "https://dev.test.com/uploads/1/210610/000918792585.jpg",
        "https://dev.test.com/uploads/1/210610/002904468645.png",
      ],
    },
  ],
]
`;

exports[`Show action 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ignoreError": true,
        "url": "/api/products/1?include=detail",
      },
    ],
    Array [
      Object {
        "url": "/api/carts/count",
      },
    ],
    Array [
      Object {
        "url": "/api/products/1/shipping-tpl?filterRulesByCity=1",
      },
    ],
    Array [
      Object {
        "data": Object {
          "quantity": 1,
          "skuId": 3,
        },
        "method": "post",
        "url": "/api/carts",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`Show basic 1`] = `
<div>
  <taro-view-core
    class="mx-page css-17ragm3-Component"
    style="--mx-colors-gray-50: #f9fafb; --mx-colors-gray-100: #f3f4f6; --mx-colors-gray-200: #e5e7eb; --mx-colors-gray-300: #d1d5db; --mx-colors-gray-400: #9ca3af; --mx-colors-gray-500: #6b7280; --mx-colors-gray-600: #4b5563; --mx-colors-gray-700: #374151; --mx-colors-gray-800: #1f2937; --mx-colors-gray-900: #111827; --mx-space-0: 0; --mx-space-1: 4; --mx-space-2: 8; --mx-space-3: 12; --mx-space-4: 16; --mx-space-5: 20; --mx-space-6: 24; --mx-space-7: 28; --mx-space-8: 32; --mx-space-9: 36; --mx-space-10: 40; --mx-space-11: 44; --mx-space-12: 48; --mx-space-14: 56; --mx-space-16: 64; --mx-space-20: 80; --mx-space-24: 96; --mx-space-28: 112; --mx-space-32: 128; --mx-space-36: 144; --mx-space-40: 160; --mx-space-44: 176; --mx-space-48: 192; --mx-space-52: 208; --mx-space-56: 224; --mx-space-60: 240; --mx-space-64: 256; --mx-space-72: 288; --mx-space-80: 320; --mx-space-96: 384; --mx-fontSizes-xs: 12; --mx-fontSizes-sm: 14; --mx-fontSizes-base: 16; --mx-fontSizes-lg: 18; --mx-fontSizes-xl: 20; --mx-fontSizes-2xl: 24; --mx-fontSizes-3xl: 30; --mx-fontSizes-4xl: 36; --mx-fontSizes-5xl: 48; --mx-fontSizes-6xl: 60; --mx-fontSizes-7xl: 72; --mx-fontSizes-8xl: 96; --mx-fontSizes-9xl: 128;"
  >
    <taro-swiper-core
      autoplay="true"
      circular="true"
      indicatoractivecolor="rgba(0, 0, 0, .3)"
      indicatorcolor="rgba(0, 0, 0, .1)"
      indicatordots="true"
      style="height: 100vw; background-color: rgb(255, 255, 255);"
    >
      <taro-swiper-item-core>
        <taro-image-core
          mode="aspectFit"
          src="https://dev.test.com/uploads/1/210610/002439539409.png"
          style="width: 100%; height: 100%;"
        />
      </taro-swiper-item-core>
      <taro-swiper-item-core>
        <taro-image-core
          mode="aspectFit"
          src="https://dev.test.com/uploads/1/210610/000924427937.jpg"
          style="width: 100%; height: 100%;"
        />
      </taro-swiper-item-core>
      <taro-swiper-item-core>
        <taro-image-core
          mode="aspectFit"
          src="https://dev.test.com/uploads/1/210610/000918792585.jpg"
          style="width: 100%; height: 100%;"
        />
      </taro-swiper-item-core>
      <taro-swiper-item-core>
        <taro-image-core
          mode="aspectFit"
          src="https://dev.test.com/uploads/1/210610/002904468645.png"
          style="width: 100%; height: 100%;"
        />
      </taro-swiper-item-core>
    </taro-swiper-core>
    <taro-view-core
      class="mx-card"
    >
      <taro-view-core
        class="mx-card-body"
      >
        <taro-view-core>
          <taro-text-core
            style="color: rgb(59, 130, 246);"
          >
            <taro-text-core>
              ￥
            </taro-text-core>
            9
          </taro-text-core>
          <taro-text-core
            style="color: rgb(107, 114, 128);"
          >
            价格:￥
            <taro-text-core
              style="text-decoration: line-through;"
            >
              10
            </taro-text-core>
          </taro-text-core>
        </taro-view-core>
        <taro-view-core
          style="flex-direction: row; justify-content: space-between; align-items: center; display: flex;"
        >
          <taro-view-core>
            <taro-view-core>
              alias commodi ipsam
            </taro-view-core>
            <taro-view-core
              style="color: rgb(107, 114, 128);"
            >
              consequatur officia sequi
            </taro-view-core>
          </taro-view-core>
          <taro-button-core
            class="mx-btn"
            opentype="share"
            style="flex-direction: column; flex-shrink: 0; color: rgb(107, 114, 128); background-color: transparent; line-height: 1; justify-content: center; align-items: center; display: flex;"
          >
            <taro-text-core
              class="mxi mxi-share"
            />
            <taro-text-core>
              分享
            </taro-text-core>
          </taro-button-core>
        </taro-view-core>
      </taro-view-core>
    </taro-view-core>
    <taro-view-core
      class="mx-card"
    >
      <taro-view-core
        class="mx-list mx-list-sm mx-list-borderless"
      >
        <taro-view-core
          class="mx-list-item mx-list-has-arrow mx-list-description-left"
        >
          <taro-view-core
            class="mx-list-title"
          >
            选择
          </taro-view-core>
          <taro-view-core
            class="mx-list-detail"
          >
            尺寸 / 颜色
          </taro-view-core>
          <taro-view-core
            class="mx-list-arrow"
          >
            <taro-text-core
              class="mxi mxi-arrow-right"
            />
          </taro-view-core>
        </taro-view-core>
        <taro-view-core
          class="mx-list-item-3 mx-list-item mx-list-description-left"
        >
          <taro-view-core
            class="mx-list-title"
          >
            发货
          </taro-view-core>
          <taro-view-core
            class="mx-list-detail"
            style="flex-direction: row; justify-content: space-between; display: flex;"
          >
            <taro-text-core>

              包邮
            </taro-text-core>
            <taro-text-core
              style="color: rgb(156, 163, 175);"
            >
              销量
              0
            </taro-text-core>
          </taro-view-core>
        </taro-view-core>
      </taro-view-core>
    </taro-view-core>
    <taro-view-core
      class="mx-divider"
    >
      商品详情
    </taro-view-core>
    <taro-view-core
      style="background-color: rgb(255, 255, 255);"
    >
      <taro-view-core
        class="mx-rich-text"
      >
        <p>
          <img
            alt="14 - 3 (2).png"
            src="https://dev.test.com/uploads/images/20210528/1622199614338752.png"
          />
        </p>
        <p>
          Alice: he had a cons
          <span
            style="color: #548DD4;"
          >
            ultation about this, and after a few minutes it puffed away without speaking, but at any r
          </span>
          ate,' sai
          <span
            style="background-color: #F79646;"
          >
            d Alice: 'allo
          </span>
          w me to introduce it.' 'I don't know what a long way. So they got.
        </p>
      </taro-view-core>
    </taro-view-core>
    <taro-view-core
      style="background-color: rgb(255, 255, 255);"
    >
      <taro-view-core
        class="mx-divider"
        style="flex-direction: row; align-items: center; display: flex;"
      >
        商品价格说明
        <taro-text-core
          class="mxi mxi-arrow-down"
        />
      </taro-view-core>
    </taro-view-core>
    <taro-view-core
      class="mx-footer-bar __mx-footer-bar-1"
    >
      <taro-button-core
        class="mx-btn mx-footer-bar-icon"
      >
        <taro-text-core
          class="mxi mxi-home"
        />
        店铺
      </taro-button-core>
      <taro-button-core
        class="mx-btn mx-footer-bar-icon"
      >
        <taro-view-core
          class="at-badge cart-num"
        >
          <taro-text-core
            class="mxi mxi-cart"
          />
          <taro-view-core
            class="at-badge__num"
          >
            1
          </taro-view-core>
        </taro-view-core>
        购物车
      </taro-button-core>
      <taro-view-core
        class="mx-btn-theme mx-btn-theme-pill"
      >
        <taro-view-core
          class="mx-btn-group"
        >
          <taro-button-core
            class="mx-btn mx-btn-secondary mx-footer-bar-btn"
          >
            加入购物车
          </taro-button-core>
          <taro-button-core
            class="mx-btn mx-btn-primary mx-footer-bar-btn"
          >
            立即购买
          </taro-button-core>
        </taro-view-core>
      </taro-view-core>
    </taro-view-core>
    <taro-view-core
      class="at-float-layout sku-picker"
    >
      <taro-view-core
        class="at-float-layout__overlay"
      />
      <taro-view-core
        class="at-float-layout__container layout"
      >
        <taro-view-core
          class="layout-body"
        >
          <taro-scroll-view-core
            class="layout-body__content"
            scrolly="true"
          >
            <taro-view-core
              style="flex-direction: row; justify-content: space-between; display: flex;"
            >
              <taro-view-core
                style="display: flex;"
              >
                <taro-image-core
                  src="https://dev.test.com/uploads/1/210610/002439539409.png"
                  style="border-radius: 4rpx; overflow: hidden;"
                />
                <taro-view-core
                  style="color: rgb(59, 130, 246); align-self: flex-end;"
                >
                  <taro-text-core>
                    ￥
                  </taro-text-core>
                  9 ~ 14

                </taro-view-core>
              </taro-view-core>
              <taro-view-core
                style="margin-top: -1rem; font-weight: 100; color: rgb(107, 114, 128);"
              >
                ×
              </taro-view-core>
            </taro-view-core>
            <taro-view-core
              style="max-height: 60vh; overflow-y: scroll;"
            >
              <taro-view-core
                class="border-b"
                style="flex-direction: column; display: flex;"
              >
                <taro-view-core
                  class="sku-picker-spec-item"
                >
                  <taro-view-core
                    class="sku-picker-spec-name"
                  >
                    尺寸
                  </taro-view-core>
                  <taro-view-core>
                    <taro-text-core
                      class="sku-picker-spec-value"
                    >
                      S
                    </taro-text-core>
                    <taro-text-core
                      class="sku-picker-spec-value"
                    >
                      M
                    </taro-text-core>
                  </taro-view-core>
                </taro-view-core>
                <taro-view-core
                  class="sku-picker-spec-item"
                >
                  <taro-view-core
                    class="sku-picker-spec-name"
                  >
                    颜色
                  </taro-view-core>
                  <taro-view-core>
                    <taro-text-core
                      class="sku-picker-spec-value"
                    >
                      红色
                    </taro-text-core>
                    <taro-text-core
                      class="sku-picker-spec-value"
                    >
                      蓝色
                    </taro-text-core>
                  </taro-view-core>
                </taro-view-core>
              </taro-view-core>
              <taro-view-core
                style="flex-direction: row; align-items: center; justify-content: space-between; display: flex;"
              >
                <taro-text-core
                  class="css-1vq2xix-SkuPicker"
                >
                  数量
                </taro-text-core>
                <taro-view-core
                  style="flex-direction: row; align-items: center; justify-content: space-between; display: flex;"
                >
                  <taro-text-core
                    style="color: rgb(107, 114, 128);"
                  >

                    剩下
                    21

                    件
                  </taro-text-core>

                  <taro-view-core
                    class="mx-stepper"
                  >
                    <taro-button-core
                      class="mx-stepper-minus"
                      disabled="true"
                    />
                    <taro-input-core
                      class="mx-stepper-input"
                      value="1"
                    />
                    <taro-button-core
                      class="mx-stepper-plus"
                    />
                  </taro-view-core>
                </taro-view-core>
              </taro-view-core>
            </taro-view-core>
            <taro-view-core
              class="mx-footer-bar __mx-footer-bar-3 __sku-picker-footer-bar-2"
            >
              <taro-view-core
                class="mx-btn-theme mx-btn-theme-pill"
              >
                <taro-view-core
                  class="mx-btn-group"
                >
                  <taro-button-core
                    class="mx-btn mx-btn-secondary mx-footer-bar-btn"
                  >
                    加入购物车
                  </taro-button-core>
                  <taro-button-core
                    class="mx-btn mx-btn-primary mx-footer-bar-btn"
                  >
                    立即购买
                  </taro-button-core>
                </taro-view-core>
              </taro-view-core>
            </taro-view-core>
          </taro-scroll-view-core>
        </taro-view-core>
      </taro-view-core>
    </taro-view-core>
  </taro-view-core>
</div>
`;

exports[`Show basic 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ignoreError": true,
        "url": "/api/products/1?include=detail",
      },
    ],
    Array [
      Object {
        "url": "/api/carts/count",
      },
    ],
    Array [
      Object {
        "url": "/api/products/1/shipping-tpl?filterRulesByCity=1",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
